### ISSUE TITLE
Emojis in Notebook Title prevent Notebook from syncing between Android and Win10

### ISSUE THREAD

**Issue by HandHsince1920** [laurent22/joplin#9219]
I use Joplin on Android and Win10. When i chose emojis in Notebook Titles it can happen, that they do not sync/are not shown on the other device (not sure if synced in the background). It seems to affect several different emojies but not all.

## Environment

Joplin version: 
Joplin 2.12.18 (prod, win32)
Sync-Version: 3
Profil-Version: 43
Unterstützter Schlüsselbund: Ja

Revision: a0a4a81

Bidirectional Links: 0.1.2
Conflict Resolution: 1.2.3
Email Note: 1.2.2
Emoji: 1.0.4
Link Graph UI: 1.5.0
OCR: 0.3.2

---

Android App Version 2.12.3 from Fdroid
Android Version 13

**Comment by mwmason** [comment]
Not sure if this is related, but I noticed that when you add an emoji to a notebook title, and the notebook has existing notes and/or sub-notebook, upon syncing, the notes/sub-notebook get deleted.

The edited notebook remains, but without the emoji.  If you add the emoji to an empty notebook, then add notes, etc... it seems to sync okay.

I am using Joplin Sync Server and Joplin Linux 2.12.19

If I should create a separate issue for this, please let me know.

Thanks!

**Comment by personalizedrefrigerator** [comment]
> Not sure if this is related, but I noticed that when you add an emoji to a notebook title, and the notebook has existing notes and/or sub-notebook, upon syncing, the notes/sub-notebook get deleted.

Additional information may be helpful here:

- Is the Emoji being added by right-clicking, selecting "Edit", then "Select Emoji", or by inserting an Emoji character into the title? (I'm assuming inserting as a character in the title.)
- Is encryption enabled?
- Are two Joplin 2.12.19 clients syncing with each other?

So far, I haven't been able to reproduce this issue with Joplin 2.13.5 and Joplin Server (I haven't tried Joplin 2.12.19 yet).

> I use Joplin on Android and Win10. When i chose emojis in Notebook Titles it can happen, that they do not sync/are not shown on the other device (not sure if synced in the background). It seems to affect several different emojies but not all.

I haven't been able to reproduce this yet &mdash; this additional information might be helpful:
- What version of Joplin is being used on Windows 10?
- What are some examples of Emojis that don't work?


Thank you for reporting this!

**Comment by mwmason** [comment]
> * Is the Emoji being added by right-clicking, selecting "Edit", then "Select Emoji", or by inserting an Emoji character into the title? (I'm assuming inserting as a character in the title.)

I used Edit, then Select Emoji.

> * Is encryption enabled?

No

> * Are two Joplin 2.12.19 clients syncing with each other?

I have a Linux Client (2.12.19) and the Android Client (2.12.3).  However, when this first occurred, I closed (forced closed) the Android Client because I suspected it could be a conflict with the two clients.  But it still occurred with just the Linux Client syncing.

A few days later, I noticed that the Android Client had a duplicate of the errant folders - one with the Emjoi, and one without.  I was able to delete the non-emoji folder and the client sync'd w/o error.  Both clients now have only the emjoi folders present.

Finally, I also tested creating an empty folder, added the emoji (Edit>Select Emoji), and then added notes.  This sync'd correctly.   My original failure had been on non-empty folders.

Thanks for looking into this

**Comment by HandHsince1920** [comment]
> > I use Joplin on Android and Win10. When i chose emojis in Notebook Titles it can happen, that they do not sync/are not shown on the other device (not sure if synced in the background). It seems to affect several different emojies but not all.
> 
> I haven't been able to reproduce this yet — this additional information might be helpful:
> 
>     * What version of Joplin is being used on Windows 10?
> 
>     * What are some examples of Emojis that don't work?
> 
> 
> Thank you for reporting this!

Hi,

Thanks for replying:

Windows:
Joplin 2.12.19 (prod, win32)
Sync-Version: 3
Profil-Version: 43
Unterstützter Schlüsselbund: Ja

Revision: 7081046

Bidirectional Links: 0.1.2
Conflict Resolution: 1.2.3
Email Note: 1.2.2
Emoji: 1.0.4
Link Graph UI: 1.5.0
OCR: 0.3.2

I unfortunately cannot remember which emojies did not work. It may have been the japan (🗾) emoji and maybe also some flags.
I also had another syncing issue recently, when I created a note from my android app. It would not sync until I had cleared its title and renamed it. I believe there must have been some special character in the title that broke the syncing. I wonder if it might be helpful to only allow simple character sets in titles and re-format/ignore special characters

**Comment by personalizedrefrigerator** [comment]
Thank you for the detailed information! I can reproduce this!

Here's what I've found so far (Joplin dev (3560bc62a27b6505dfa41ae96d313fa592b9463f) on Ubuntu 23.10 $\to$ Joplin Android, same dev commit):
- A notebook that existed before syncing with Joplin Server for the first time (notebook emoji: `🤔` title: `Welcome! 🦑`) doesn't show up after initial sync.
    - The notebook has several notes (mostly the welcome notes, one test HTML note)
- Creating a new notebook with a single note and **with the same title and notebook emoji** syncs fine
- Editing a note in the notebook that doesn't sync makes the note appear under "All notes" on Android. The container note is listed as `...`.
- If I export the original notebook as JEX, re-import it, then sync, the imported JEX notebook syncs fine.

Log: [android-log.txt](https://github.com/laurent22/joplin/files/13401259/android-log.txt)

The issue seems to be related to the initial sync.

**Comment by personalizedrefrigerator** [comment]
On Android, clearing data & redownloading from the sync target ([requires patch](https://github.com/laurent22/joplin/compare/3560bc62a27b6505dfa41ae96d313fa592b9463f...personalizedrefrigerator:joplin:7ea3e1c0ed9c68496dc2a70896935b4b4834e812)) doesn't download notebooks with emojis that synced previously.

Before doing this, I created two notebooks:
- `Test emoji 🤔🦑` (no notebook emoji)
- `Test emoji 🦑` (notebook emoji: `🤔`)

Neither showed up on the initial sync of the Android device, but pressing sync again caused both to show up.

Next, clicking  "clear remote data and re-upload to the sync target" on the desktop app, re-syncing, then clicking "delete remote data and re-download from sync target" on mobile (again) syncs most of the notebooks that didn't show up previously. However, the original notebook created above (notebook emoji: `🤔` title: `Welcome! 🦑`) doesn't sync, though its copy (notebook emoji: `🤔` title: `Welcome (1)! 🦑`) and other notebooks with similar titles do.

I have note history turned on. **Edit**: If I disable note history, click "clear remote data and re-upload to the sync target" on desktop, and re-download from the sync target on mobile, everything seems to sync.

**Comment by personalizedrefrigerator** [comment]
I'm also able to reproduce this syncing two Joplin desktop clients.

The issue might be related to Joplin Server/Cloud's `delta` functionality.

I'm able to find the ID of a notebook that isn't syncing in a list of all notes/notebooks/etc on the server. However, the result of a `delta` doesn't include this ID, while it does seem to include the IDs of notebooks/notes that are syncing.

I've been testing this with the following script in Joplin's development tools:
```js
// Here, I'm using "var" to allow running the same code multiple times (without "can't redeclare"
// errors)

// Get a reference to the sync target
const { reg } = require('@joplin/lib/registry.js');
var sync = reg.syncTarget();
var fileApi = await sync.fileApi(); // Joplin Server file API

// The ID of a notebook that isn't syncing locally
var notSyncingId = '8c4220088e7e46c089665a65a94f8275';

// Get a list of all items on the sync target that match the notebook that isn't syncing
var matches = []
var list = undefined;
for (let page = 0 ; page < 100 && (list?.hasMore ?? true); page++) {
    list = await fileApi.list('', { page, context: list?.context });
    matches.push(...list.items.filter(item => item.path.includes(notSyncingId)));
}
// Log the matching items
console.log('matches', matches);

// Do a delta with nothing. This should (maybe? I'm unsure here...) return all items that would
// be synced in an an initial sync.
var delta = await fileApi.delta('');
console.log('delta (with nothing) has ', delta.items.length, 'items');
var deltaMatches = delta.items.filter(item => item.path.includes(notSyncingId))
console.log('items from delta matching the non-syncing notebook', deltaMatches)

// Look for items that have synced successfully in the delta. Interestingly, 
// the last item, a folder, does **not** appear in the output.
var areSyncing = [ '5e56ce958e35464db49ddcebbe6a7e91', '1a5eba7cfc2140c9a9886034c1da4ed4', '683efb37e0fe43e1a74f6b98d907fc55', '39070855b97744b890f9c71233ae1fca' ];
var deltaMatches2 = delta.items.filter(item => {
    for (const isSyncingId of areSyncing) {
        if (item.path.includes(isSyncingId)) {
            return true;
        }
    }
    return false;
});
console.log('items from delta matching a syncing item', deltaMatches2)
```

Output:
```
FileApi: list 
FileApi: list 
FileApi: list 
FileApi: list 
FileApi: list 
FileApi: list 
FileApi: list 
FileApi: list 
VM5256:12 matches [{…}]
    0: {path: '8c4220088e7e46c089665a65a94f8275.md', updated_time: 1700345319365, jop_updated_time: undefined, isDir: false, isDeleted: false}
    length: 1
/home/builder/Documents/joplin2/packages/utils/dist/Logger.js:213 15:23:11: FileApi: delta 
VM5256:15 delta (with nothing) has  97 items
VM5256:17 items from delta matching the non-syncing notebook []
VM5256:28 items from delta matching a syncing item
    (3) [{…}, {…}, {…}]
        0: {path: '1a5eba7cfc2140c9a9886034c1da4ed4.md', updated_time: 1700344156719, jop_updated_time: 1700344055224, isDir: false, isDeleted: false}
        1: {path: '683efb37e0fe43e1a74f6b98d907fc55.md', updated_time: 1700344167469, jop_updated_time: 1700349703422, isDir: false, isDeleted: false}
        2: {path: '5e56ce958e35464db49ddcebbe6a7e91.md', updated_time: 1700345341194, jop_updated_time: 1700345335993, isDir: false, isDeleted: false}
        length: 3
```

@laurent22 Is this expected behavior for `delta`?

**Comment by laurent22** [comment]
@personalizedrefrigerator, yes it should indeed return everything with `fileApi.delta('')`. Do you have a setup where you can consistently get this delta to not return that non-syncing notebook?

If so could you try print the query in `ChangeModel::changesForUserQuery`? You would do that by logging `query.toQuery()`. And from that could you please check in the Joplin Server database the non-syncing notebook and see why it wouldn't be fetched by this query? You can find the notebook using the "jop_id" parameter or the name

**Comment by HandHsince1920** [comment]
Hi,
I do not think it if related to the cloud server because i never used the cloud server. I forgot to mention, sorry(!) , but I sync to file system and this is connected via syncthing.

**Comment by personalizedrefrigerator** [comment]
> @personalizedrefrigerator, yes it should indeed return everything with `fileApi.delta('')`. Do you have a setup where you can consistently get this delta to not return that non-syncing notebook?
> 
> If so could you try print the query in `ChangeModel::changesForUserQuery`? You would do that by logging `query.toQuery()`. And from that could you please check in the Joplin Server database the non-syncing notebook and see why it wouldn't be fetched by this query? You can find the notebook using the "jop_id" parameter or the name

I haven't managed to reproduce this after pulling the recent change to Joplin server. I suspect that the issue I was experiencing was related to only a single page of items (30-100?) syncing each time I pressed "sync". My test notebooks had a large number of notes and revisions.

I'm testing sync between three clients (desktop -> Android and another desktop client).

This is a subset of the notebook titles I'm using:
<img src="https://github.com/laurent22/joplin/assets/46334387/fa4708a7-6510-430f-97c4-105cbc7733ef" width="400" alt="⏰;Scheduling;135;📑;Homework and Projects;74;😱;Security;1;⏰;Test;7;🚧;🕳️ Test‼️;80;⏰;TestSubnotebook;7;📪;𝔞_𝔱𝔢𝔰𝔱;20;Welcome!;5;Welcome!;5;Test;4;Test;1;test;5;🤔;Test 1;14;Test 1234;1;🤔;Welcome! 🦑;7;Welcome!☄️☄️☄️;8;␣‼️;8;🕝🕳️;1;🗾 Test 🗾;1;ℹ️;🧱 🛈 Welcome! 🛈 🌋;10;🪐;"/>


> Hi,
> I do not think it if related to the cloud server because i never used the cloud server. I forgot to mention, sorry(!) , but I sync to file system and this is connected via syncthing.

Thank you for providing this information!

In that case, there are probably two (or more) issues being tracked here. The issue I was experiencing (and perhaps the issue described by @mwmason) seems to be fixed by 87aeffa160db39a298aceb758861bb805a03f7c3 (though I did clear everything and restore from a backup during testing, after doing tests related to the Android device sync).

Another similar issue was [reported on the Joplin forum](https://discourse.joplinapp.org/t/odd-sync-issue/33734/20), but that issue does not seem to be fixed by 87aeffa160db39a298aceb758861bb805a03f7c3.

**Comment by personalizedrefrigerator** [comment]
I haven't managed to reproduce this &mdash; I've re-tested with Syncthing (OpenSUSE 20240130 $\leftrightarrow$ Android 7) and encryption both enabled and disabled.

I originally tested with encryption enabled and used 🕳️ and 🗾 as notebook emojis. ("🕳️" was selected because it is a multi-byte emoji, though so is "🗾"). I then tested with several 🕳️ characters in a new notebook title. All synced successfully.

I then disabled encryption and created notebooks with flag emojis (and one with the 🗾 emoji), all seem to sync successfully.

The original bug report used Android 13, Windows 10, and Joplin 2.12.3, which may be related.
